### ISSUE TITLE
Mode collapse in Conditional GAN Sequence (text) generation

### ISSUE THREAD

**Issue by rafaelvalle** [martinarjovsky/WassersteinGAN#14]
The _Wasserstein GAN_ I'm training is presenting mode collapse behavior:
Parameters and example of generated text after the 10 epochs, where each epoch has 100*128 examples, are below.

The critic loss graph is decreasing but showing oscillations.
Suggestions would be very welcome!
Namespace(boolean=0, bs=128, clip=0.01, epoch_size=100, lr=5e-05, n_epochs=100)
Dataset shape (129822,)
Building model and compiling functions...
('Generator output:', (None, 1, 128, 128))
('Critic output:', (None, 1))

```
0, (aueldablecarbld damubkeckecolait astir thin in bowpbor siry le ty therandurcing day anat yale beain ghckvincqundg"bdxk'ntqxw8v'
1,  bueldeblecarbcdsdamuqfackeckbalt astar than in tiwpgor sury ye th thetandurting kellanat lale beain ghapvincquod)"bdak'nthxwv'
2,  bueldafredawuckodelreficha kbalv astar ;hathin  o0 wor cure ey rh th stodeutine sellavet lale  eain chapvinckunvewb'az'xthg\d
3, (bueldablecarbcdndamuqlabkecobalv artar , an in tin bor ciry ef rh thettndluting dellavat lale  eain jhapvincquid)wb'av'nthxlDy'
4,  bumldeblecawbcdsdaluqfackackbalt astar that in tow gor cure ey rh thestndurting kellanet lale  eain ghapvinckuod)"b'sz'xthxw(y
5,  bueldabrecawuclodeldadbchanksaltursdarithit os tmn #39;s re lerperyo raouetcane key tne. <ate begad ghakmfgelunc)"bdayex hgr8v'
6, raiedlibldrisblx grjucambngcoln-tursdiait in os worg ba sicr he pe co ma-iuld berday tia. .ate began ghcuzid qunkg"zlxk'n wejxvd
7,  bueldifredawuckopelredicha qualt rsdarithathon ton toxZs(xelyerre to saouestene sel ane? late  eaid chapadgelunve(bdaz'x hxr8v
8, pbceldefrldrwuqlefetdadibh, qual; rsdar tsathis, or wor ct	wl f re to stouluteve sevltve? late  eaid chapadgckjnve(bzaz'x hg\d
9,  bumldabkecawbcdsdalumfichacksait astarithin in tonpgor s rellertertherandetcing key anet <ate beain ghakmidclund)"bdak'xtqxr8v'
10, rauedjabldrispld dajucllongcolait artir t an in win bor sicg lf th themandlrning day aiat yale beain jhckvincqvidg"zlxk'n wxwDv'
```

**Comment by martinarjovsky** [comment]
Interesting. A lot of weird things can happen when switching to rnns. Can I take a look at the architecture and training curves? Are you using batchnorm / layer normalization?

**Comment by rafaelvalle** [comment]
Yes, but note that I using CNNs for sequence prediction.
I'm running this file https://github.com/rafaelvalle/neural_network_control_improvisation/blob/master/wcgan_text.py and the models are in https://github.com/rafaelvalle/neural_network_control_improvisation/blob/master/models.py

Critic is a Crepe CNN and the generator is a simple architecture with 2 transposed convs.
Thanks for the help!

**Comment by rafaelvalle** [comment]
Generator
```
        # fully-connected layer
        layer = batch_norm(DenseLayer(layer, 1024))
        # fully-connected layer
        layer = batch_norm(DenseLayer(layer, 1024))
        # project and reshape
        layer = batch_norm(DenseLayer(layer, 128*34*34))
        layer = ReshapeLayer(layer, ([0], 128, 34, 34))
        # two fractional-stride convolutions
        layer = batch_norm(Deconv2DLayer(
            layer, 128, 5, stride=2, crop='same', b=None, nonlinearity=lrelu))
        layer = Deconv2DLayer(
            layer, 1, 6, stride=2, crop='full', b=None,
            nonlinearity=tanh_temperature)
```

Critic
```
        # CREPE
        # form words from sequence of characters
        layer = batch_norm(Conv2DLayer(layer, 1024, (7, 128), nonlinearity=lrelu))
        layer = MaxPool2DLayer(layer, (3, 1))
        # temporal convolution, 7-gram
        layer = batch_norm(Conv2DLayer(layer, 512, (7, 1), nonlinearity=lrelu))
        layer = MaxPool2DLayer(layer, (3, 1))
        # temporal convolution, 3-gram
        layer = Conv2DLayer(layer, 256, (3, 1), nonlinearity=lrelu)
        layer = Conv2DLayer(layer, 256, (3, 1), nonlinearity=lrelu)
        layer = Conv2DLayer(layer, 256, (3, 1), nonlinearity=lrelu)
        layer = Conv2DLayer(layer, 256, (3, 1), nonlinearity=lrelu)
        layer = flatten(layer)
        # fully-connected layers
        layer = dropout(DenseLayer(layer, 1024, nonlinearity=rectify))
        layer = dropout(DenseLayer(layer, 1024, nonlinearity=rectify))
```

**Comment by rafaelvalle** [comment]
I'm currently re-running with batchnorm and lower learning rate.
![g_updates30](https://cloud.githubusercontent.com/assets/2302607/23374829/c833a6f2-fcda-11e6-9d93-d5ba315dadc6.png)

**Comment by rafaelvalle** [comment]
I suspect that the generator is ignoring the noise input and relying on conditions to generate output. I'll modify the model and post the results.

**Comment by NickShahML** [comment]
@rafaelvalle I've experimented with wgan's and text generation myself. For me, the discriminator stays at a loss of -8.2 and the generator stays at a loss of -0.00012 (it changes a little bit but stays in that range). I think the discriminator is getting stuck, which seems to be your problem as well. I've tried annealing the learn rates but didn't have an effect.

I'm considering different discriminator architectures and I'm convinced that they need to be (RNN's have a really hard time with gradients and don't converge for me).

Also how are you actually generating your characters? I haven't had a chance to look at your code, but do you generate softmaxed probability distributions and and pass those to the discriminator?

**Comment by rafaelvalle** [comment]
@NickShahML I think our problems are different. I've had generators able to produce word that have high probability in my data. However, the generator has one single mode for each label and I suspect this is because the generator is ignoring the noise input and relying on the labels only:  I concatenate the noise and the labels at the first layers. I'm slightly changing the architecture such that the noise goes through a fully connected dense layer before being concatenated with the labels.

On another note, it really impresses me that the model does not converge if the discriminator has batchnorm.

**Comment by NickShahML** [comment]
@rafaelvalle if I use a generator RNN without batch norm on the critic the model does not converge. The discriminator basically overpower's the generator. So in other words, I have to use batch norm on the critic to get anywhere at all. You may be right: You have to use a convolutional generator to actually reach convergence.

**Comment by martinarjovsky** [comment]
I will have a deeper look at all of this when I have more time next week, but the general observation from the curves is that the critic is not well trained till optimality ( - Loss of the critic should never be negative, since outputing 0 would yeald a better loss so this is a huge red flag). 

@NickShahML there is no overpowering. What you are seeing is that without batchnorm on the critic the critic is not trained till optimality, therefore the loss ceases to be a good wasserstein estimate, the numbers mean very little and the updates to the generator are crap.

For now batchnorm seems to be important on the critic to get it close to optimality in a few iterations. We've stressed this in other issues as well.

**Comment by NickShahML** [comment]
@martinarjovsky  Thank you for that clarification. I didn't realize that negative losses in the critic are a huge red flag. Can you clarify what this indicates? I'll investigate why this is occurring.

If the generator loss is _also_ negative, is that a red flag? I would imagine that a discriminator raw logit could be negative or positive (since we are clipping `[-0.01, 0.01]`). Thus the loss of the generator could potentially be negative. 

`self._g._loss = tf.reduce_mean(d._fake_logits_evaluation)`

The raw outputs of the discriminator can be negative meaning that the generator's loss could be negative itself. Let me know if this is incorrect. Really appreciate your help!

**Comment by rafaelvalle** [comment]
Here are some plots of the generated data with and without batch norm on the Crepe critic.
Note that **without** batch norm, there is some change in the generator, suggesting that it's learning something; **with** batch norm there's little to no change in the output of the generator.

Loss Curve without Batch Norm
![Loss Curve Batch Norm](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/nobatch_g_updates20.png)
Generator Output, **without** Batch Norm 0th, 5th and 10th iterations
![No batch 0th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/nobatch_0.jpg)
![No batch norm 5th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/nobatch_5.png)
![No batch 10th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/nobatch_10.png)

Loss Curve with Batch Norm
![Loss Curve with Batch Norm](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/batch_g_updates20.png)
Generator Output, **with** Batch Norm 0th, 5th and 10th iterations
![Batch Norm 0th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/batch_0.png)
![Batch Norm 5th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/batch_5.png)
![Batch Norm 5th iteration](https://raw.githubusercontent.com/rafaelvalle/neural_network_control_improvisation/master/batch_10.png)

**Comment by NickShahML** [comment]
@rafaelvalle  can you quickly report what the generator and discriminator architectures were? RNN or CNN or Dense?

**Comment by rafaelvalle** [comment]
Both critic and generator are CNN. I asked Xiang Zhang, the main author on the Crepe Network paper, about batch norm in Crepe networks. He explained that he zero-padded the data such that it would fit the specs of the input layer of the models he was training. This zero-padding makes the last features in length dimension have close to zero variance, which makes the gradients blow when backpropagated. 
Does anyone think that noise-padding instead of zero-padding would allow the use of batch-norm and not significantly perturb training?

**Comment by NickShahML** [comment]
@rafaelvalle  I understand that you're trying to make this WGAN work but why are you trying to make the Crepe CNN work? Sure it may have some great features, but _first_ try to make a regular convolutional network work first for your task. If that works, _then_ I would try Crepe Networks. There's also different architectures to explore such as bytenet and dense CNN nets.

**Comment by rafaelvalle** [comment]
Temporal CNNs, of which the Crepe network is a type, have been used for text encoding. 
I'm comparing the performance of different pairs of critic/generator networks: naive CNN, DCGAN, Crepe, LSTM, CNN-RNN...

Learning Deep Representations of Fine-Grained Visual Descriptions
https://arxiv.org/pdf/1605.05396.pdf

Generative Adversarial Text to Image Synthesis
https://arxiv.org/pdf/1605.05396.pdf

**Comment by omair-kg** [comment]
@martinarjovsky hi, you mentioned above that the loss for the critic should never be negative. In case it is going to negative, what is the counter measure? Increase the iterations for the critic?
